TACC/Core-CMS
# feat: tup 706 conditionally serve raw markup

## Overview

**Goal:** Render TACC News outside of TACC (when we should not just link the user).

> [!IMPORTANT]
> Supports requirement for LCCF website to show LCCF news from TACC.

<details><summary><strong>Feature:</strong> Allow client to load raw page content markup (no header, no footer).</summary>

<br />

**Use Case?** Load news pages of one website (e.g. TACC) on other websites (e.g. LCCF, WeTeach CS). By having raw markup available at a URL, another website can load and inject that markup on its own page, thus borrowing news articles and lists rather than duplicating them.

**Proof of Concept?** In #868, I create and run an app that injects remote content. (I am still thinking how/when best to run the app.)

</details>

<details><summary><strong>Bonus:</strong> Allow dev to render a page with any Core template (or one of their own).</summary>

<br />

**Use Case?** Test new templates for future custom CMS features without editing Core.

**How to Use One of Their Own?** Add (or override) a template into `taccsite_cms/templates` via Dockerfile, as is done in some Core-CMS-Custom projects ([apcd_cms](https://github.com/TACC/Core-CMS-Custom/tree/main/apcd_cms/src/taccsite_cms/templates), [ecep_cms](https://github.com/TACC/Core-CMS-Custom/tree/main/ecep_cms/src/taccsite_cms/templates)) and TUP ([tup-cms](https://github.com/TACC/tup-ui/tree/main/apps/tup-cms/src/taccsite_cms/templates)).

</details>

## Related

- solution for [TUP-706](https://tacc-main.atlassian.net/browse/TUP-706)
- supports #876
- required by #868
- requires #871

## Changes

- **added** middleware to change template at runtime before render

## Testing

0. In `custom_app_settings.py`, add new middleware to `CUSTOM_MIDDLEWARE` e.g.
    ```py
    CUSTOM_MIDDLEWARE = ['taccsite_cms.middleware.cms_template.CMSTemplateMiddleware']
    ```
2. Load pages with **recognized** query parameters.
    | `?raw` | `template=raw.html` |
    | - | - |
    | <img width="920" alt="good - raw" src="https://github.com/user-attachments/assets/67e5cea0-7ba8-497a-8bca-ddc88bceb92a"> | <img width="920" alt="good - template raw html" src="https://github.com/user-attachments/assets/ad74202a-64de-4f50-ba42-bd6777f154df"> |
3. Verify page loads with `raw.html` template.
4. Load pages with **ineffectual** query parameters.
    | (no query) | `?rawng` | `template=raw` | `template=raw` |
    | - | - | - | - |
    | <img width="920" alt="noop " src="https://github.com/user-attachments/assets/576533e9-97fb-4eff-b710-7841514c0465"> | <img width="920" alt="noop - rawng" src="https://github.com/user-attachments/assets/3291e5f1-7cd1-4e86-8164-48767119c8e2"> | <img width="920" alt="noop - template" src="https://github.com/user-attachments/assets/5e2a7a0a-cbe4-4f67-b7dc-d68f1a9ab366"> | <img width="920" alt="noop - template raw" src="https://github.com/user-attachments/assets/6e93c7af-78c6-4b09-9edb-d19e74e2c17b"> |

5. Verify page loads with template that the page would normally use.
